### PR TITLE
Updating deployment process for better external use of the project

### LINKED ISSUE #449: Create README on how external projects should depend/migrate on dex-contracts
Follow up task from https://github.com/gnosis/dex-contracts/pull/400

> Can you give the exact steps, for a brand new external project, on how to migrate? I see there's several options, but I'm not 100% sure about the one you are going for. A readme with steps would be ideal.
Also, what I'm more interested, what is the official way to depend on this contracts. Create a solidity contract that uses the contracts.

### PR DESCRIPTION
**Pull request by josojo**
This PR modifies the way artifacts are handled in the migration process. This enables the batch exchange deployment from external projects, without publishing the contracts in the npm package.

closes: #449 

An example of how this repo can be used for an external project can be found here:
https://github.com/gnosis/dex-contracts-integration-example

---
testplan:
Run the commands from the readme on 
https://github.com/gnosis/dex-contracts-integration-example

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@gnosis.pm/dex-contracts",
-  "version": "0.1.3",
+  "version": "0.1.4",
```
Comment (fleupold) [inline package.json]: Should the version bump be its own commit?

File: src/migration/utilities.js [hunk 2]
```diff
@@ -1,14 +1,32 @@
-function getDependency(artifacts, network, deployer, path) {
+function initializeContract(path, deployer, accounts) {
+  const contract = require("truffle-contract")
+
+  const Contract = contract(require(path))
```
Comment (fleupold) [inline src/migration/utilities.js]: Is there a reason we name instance with uppercase and the imported libraries wit lower-case? I was under the assumption we usually do it the other way around.

File: src/migration/utilities.js [hunk 3]
```diff
@@ -1,14 +1,32 @@
...
+  Contract.setNetwork(deployer.network_id)
+  // For some reason the automatic value calculation is not working, hence we do:
+  Contract.defaults({
+    from: accounts[0],
+    gas: 8e6,
+  })
+  return Contract
+}
+
+function getDependency(artifacts, network, deployer, accounts, path) {
```
Comment (fleupold) [inline src/migration/utilities.js]: Should we just pass a single account here? It doesn't seem like we need all.

File: src/migration/utilities.js [hunk 4]
```diff
@@ -1,14 +1,32 @@
...
   let Contract
 
   if (isDevelopmentNetwork(network)) {
-    Contract = artifacts.require(path)
+    // If this migration script is used from the repository dex-contracts, the contract
+    // data is received via the artificats.require.
+    // If this migration script is used from an external project, the first try statement
+    // will fail and it will get the contracts from the function initializeContract.
+    try {
+      Contract = artifacts.require(path.split("/").pop())
```
Comment (fleupold) [inline src/migration/utilities.js]: Would it be possible to always load the resource from disk? What's the advantage of using the artefact when we are on development?

File: src/migration/utilities.js [hunk 5]
```diff
@@ -1,14 +1,32 @@
...
+function initializeContract(path, deployer, accounts) {
+  const contract = require("truffle-contract")
+
+  const Contract = contract(require(path))
+  Contract.setProvider(deployer.provider)
+  Contract.setNetwork(deployer.network_id)
+  // For some reason the automatic value calculation is not working, hence we do:
+  Contract.defaults({
+    from: accounts[0],
+    gas: 8e6,
```
Comment (fleupold) [inline src/migration/utilities.js]: Is this for deployment? We need to be a bit careful as this exceeds the default ganache limit. Would 6M be sufficient for deployment?

**Review by fleupold — APPROVED** [review]
Smallish comments.

File: src/migration/utilities.js (hunk 5 above)
Comment (josojo) [reply src/migration/utilities.js]: Yeah, it is for deployment. I can change it for 6.5M gas. 
Generally, I am very unsatisfied that we don't get these numbers automatically if we migrate BatchExchange as a dependency

File: src/migration/utilities.js (hunk 4 above)
Comment (josojo) [reply src/migration/utilities.js]: If this script should be used in dex-contracts, then we don't have the artifact in the path provided. Then we need to get it out of the build folder.

File: src/migration/utilities.js [hunk 6]
```diff
@@ -1,14 +1,32 @@
-function getDependency(artifacts, network, deployer, path) {
+function initializeContract(path, deployer, accounts) {
+  const contract = require("truffle-contract")
```
Comment (nlordell) [inline src/migration/utilities.js]: I think we might need to add `truffle-contract` to the dependencies or peerDependencies? That might be a "heavy" dependency that we were trying to avoid. @anxolin what do you think?

File: src/migration/utilities.js (hunk 4 above)
Comment (nlordell) [reply src/migration/utilities.js]: ~~I'm wondering when we deploy, will the artifact networks get updated when we don't use `artifacts`?~~
edit: I see it prefers Truffle `artifact` so this shouldn't be a problem.

**Review by nlordell — COMMENTED** [review]
I like the changes, but I wonder if we have to fix our dependencies to make this work. We might have to include the solidity libraries as dependencies as well as truffle contract.

**Review by bh2smith — APPROVED** [review]
Looks pretty good. I wouldn't mind seeing some comments in the scripts describing what is happening in some places.

File: package.json (hunk 1 above)
Comment (nlordell) [reply package.json]: I also notice that `0.1.4` tag was pushed and auto-deployed to NPM. I think we should include a check deploy will fail if the tag is not on `master`.

Should I yank the version from NPM?

**Comment by bh2smith** [comment]
I have tried to run the test plan on the example repo as follows;

- Clone the integration repo
- run the commands from the readme

And have run into several problems:

missing truffle dependency, no mention of need for ganache in readme, some other missing dependencies required by truffle-config (i.e. `dotenv`). I am have reported a few issues in the new repo and still working towards making this work. Will report back with, hopefully, some answers soon.

File: package.json (hunk 1 above)
Comment (josojo) [reply package.json]: How would I test the deployment scripts for 
https://github.com/gnosis/dex-contracts-integration-example, I need the package in there... Having it published only locally makes it testable for me, but not for the others.

Should I do a manual 0.1.4-dev npm release?

File: src/migration/utilities.js (hunk 6 above)
Comment (josojo) [reply src/migration/utilities.js]: Yeah, I tried to avoid it and put it all just in the npm package.json of 
https://github.com/gnosis/dex-contracts-integration-example
While this is not "very clean", it keeps the size down of dex-contracts npm package.

For which applications is the small size important?

**Review by fedgiac — APPROVED** [review]
I can't compile [dex-contracts-integration-example](https://github.com/gnosis/dex-contracts-integration-example), otherwise it looks good.

File: src/migration/utilities.js (hunk 6 above)
Comment (anxolin) [reply src/migration/utilities.js]: Yes, agree with Nick. 

@josojo , super nice that you are doing an example on how to deploy the contracts from another project.

I expect that many projects won't even do this, and they will just use an interface or the original contract and use the rinkeby/mainnet deployments. Sometimes that approach is more straightforward.

Alex, for https://github.com/gnosis/dex-contracts-integration-example I think it would be better if you define the steps in the README to add it to any project.

```
npm install dex-contracts
npm install <dependencies like truffle>
...
```

This way is more explicit what we depend on.

WDYT?

Comment (josojo) [reply src/migration/utilities.js]: I disagree.
For every other project, owl-token, solidity-datastructures, etc, we always create the npm modules, such that later on others can easily build on them.
And we benefit from that a lot. For example, imagine the OWL contracts, all libraries, and all the other things are would be in this dex-contracts project. Then it would be harder to navigate for sure.

Comment (josojo) [reply src/migration/utilities.js]: >I expect that many projects won't even do this, and they will just use an interface or the original contract and use the rinkeby/mainnet deployments. Sometimes that approach is more straightforward.

If you wanna write any meaningful tests, then you need to have the deployment scripts etc...

Comment (anxolin) [reply src/migration/utilities.js]: I've got a bit lost. 

My understanding is that we want to decide if we add the dependencies: For example `truffle-contracts` in our dependencies.

I would say to not do this, even as peerDependencies. A simplistic integration might be interested ONLY in the networks.json and the ABI, so we don't want to bring all the dependencies.

This is why, I thought of adding additional manual npm install. An alternative would be to create another npm-package or using yarn workspaces to create a package that depends on `dex-contracts` and all the other stuff

Comment (anxolin) [reply src/migration/utilities.js]: > If you wanna write any meaningful tests, then you need to have the deployment scripts etc..

I agree, for some people. 
Other people can be happy with `dex-contracts` tests already (nice job), so in their project, they just mock dFusion contracts, and test their own and the interactions with the mocked dFusion contracts

Comment (josojo) [reply src/migration/utilities.js]: Okay, @anxolin :) then we are on the same page. Sorry, I was also a little bit confused. But actually we were in favor of the same thing!

I will not list all the dependencies in a readme in https://github.com/gnosis/dex-contracts-integration-example, they just go into the package.json of the dex-contracts-integration-example.

File: package.json [hunk 7]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@gnosis.pm/dex-contracts",
-  "version": "0.1.3",
+  "version": "0.1.5-dev",
```
Comment (fleupold) [inline package.json:3]: It's pretty annoying having to publish versions to test this PR that are then public for the rest of time. Can we not use something like https://stackoverflow.com/questions/15806241/how-to-specify-local-modules-as-npm-package-dependencies to test this locally without publicly pushing new versions?

Can we please revert and delete those releases?
